File: android_cursor_grok/data/crypto/src/main/kotlin/com/transcripto/crypto/XorCipher.kt
package com.transcripto.crypto

import com.transcripto.domain.CipherMethod
import java.util.Base64
import java.nio.charset.StandardCharsets

class XorCipher : CipherMethod {
    override fun encrypt(input: String, params: Map<String, Any>): String {
        val key = (params["key"] as? String ?: "") + (params["salt"] as? String ?: "")
        if (key.isEmpty()) throw IllegalArgumentException("Clave requerida")
        val inputBytes = input.toByteArray(StandardCharsets.UTF_8)
        val keyBytes = key.toByteArray(StandardCharsets.UTF_8)
        val result = xorBytes(inputBytes, keyBytes)
        return Base64.getEncoder().encodeToString(result)
    }

    override fun decrypt(input: String, params: Map<String, Any>): String {
        val key = (params["key"] as? String ?: "") + (params["salt"] as? String ?: "")
        if (key.isEmpty()) throw IllegalArgumentException("Clave requerida")
        val inputBytes = Base64.getDecoder().decode(input)
        val keyBytes = key.toByteArray(StandardCharsets.UTF_8)
        val result = xorBytes(inputBytes, keyBytes)
        return result.toString(StandardCharsets.UTF_8)
    }

    private fun xorBytes(data: ByteArray, key: ByteArray): ByteArray {
        val result = ByteArray(data.size)
        for (i in data.indices) {
            result[i] = (data[i].toInt() xor key[i % key.size].toInt()).toByte()
        }
        return result
    }
}

File: android_flutter/transcripto/packages/data_crypto/lib/src/xor_cipher.dart
import 'dart:convert';

import 'package:domain/domain.dart';

class XorCipher implements CipherMethod {
  List<int> _effectiveKeyBytes(CipherParams p) {
    var key = p.key ?? '';
    if (p.useSalt && p.salt != null) {
      key = key + utf8.decode(p.salt!);
    }
    final kb = utf8.encode(key);
    if (kb.isEmpty) throw ArgumentError('Clave inválida');
    return kb;
  }

  @override
  String encrypt(String input, CipherParams params) {
    final keyBytes = _effectiveKeyBytes(params);
    final inputBytes = utf8.encode(input);
    final outBytes = _xorBytes(inputBytes, keyBytes);
    return base64Encode(outBytes);
  }

  @override
  String decrypt(String input, CipherParams params) {
    final keyBytes = _effectiveKeyBytes(params);
    final payloadBytes = base64Decode(input);
    final outBytes = _xorBytes(payloadBytes, keyBytes);
    return utf8.decode(outBytes);
  }

  List<int> _xorBytes(List<int> data, List<int> key) {
    final out = List<int>.generate(data.length, (i) => data[i] ^ key[i % key.length]);
    return out;
  }
}
